pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,354 @@
+	clusters, err := c.clusterCache.GetByIndex(isCKSIndexer, "true")
+	if err != nil {
+		logrus.Error("error trying to refresh EKS clusters' upstream states")
+		return
+	}
+
+	for _, cluster := range clusters {
+		if _, err := c.refreshClusterUpstreamSpec(cluster); err != nil {
+			logrus.Errorf("error refreshing EKS cluster [%s] upstream state", cluster.Name)
+		}
+	}
+}
+
+func (c *clusterRefreshController) refreshClusterUpstreamSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if cluster.Spec.AKSConfig == nil && cluster.Spec.EKSConfig == nil {
+		return cluster, nil
+	}
+
+	logrus.Infof("checking cluster [%s] upstream state for changes", cluster.Name)
+
+	if cluster.Status.AKSStatus.UpstreamSpec == nil || cluster.Status.EKSStatus.UpstreamSpec == nil {
+		logrus.Infof("initial upstream spec for cluster [%s] has not been set by eks cluster handler yet, skipping", cluster.Name)
+		return cluster, nil
+	}
+
+	// In this call, it is possible to get errors back with non-nil upstreamSpec.
+	// If upstreamSpec is nil then the syncing failed for some reason. This is reported to the user, and this function returns at the end of this if-statement.
+	// If upstreamSpec is non-nil then the syncing occurred as expected, but the node groups have health issues that are reported to the user.
+	// In this second case, the message is set on the Updated condition, but execution continues because the sync was successful.
+	upstreamConfig, err := GetComparableUpstreamSpec(c.secretsCache, cluster)
+	if err != nil {
+		var statusErr error
+		var syncFailed string
+		if upstreamConfig == nil {
+			syncFailed = ": syncing failed"
+		}
+		cluster = cluster.DeepCopy()
+		apimgmtv3.ClusterConditionUpdated.False(cluster)
+		apimgmtv3.ClusterConditionUpdated.Message(cluster, fmt.Sprintf("[Syncing error%s] %s", syncFailed, err.Error()))
+
+		cluster, statusErr = c.clusterClient.Update(cluster)
+		if statusErr != nil {
+			return cluster, statusErr
+		}
+	} else if strings.Contains(apimgmtv3.ClusterConditionUpdated.GetMessage(cluster), "[Syncing error") {
+		cluster = cluster.DeepCopy()
+		apimgmtv3.ClusterConditionUpdated.True(cluster)
+		apimgmtv3.ClusterConditionUpdated.Message(cluster, "")
+		cluster, err = c.clusterClient.Update(cluster)
+		if err != nil {
+			return cluster, err
+		}
+	}
+	if upstreamConfig == nil {
+		return cluster, err
+	}
+
+	switch {
+	case cluster.Spec.AKSConfig != nil:
+		if !reflect.DeepEqual(cluster.Status.AKSStatus.UpstreamSpec, upstreamConfig.aksConfig) {
+			logrus.Infof("updating cluster [%s], upstream change detected", cluster.Name)
+			cluster = cluster.DeepCopy()
+			cluster.Status.AKSStatus.UpstreamSpec = upstreamConfig.aksConfig
+			cluster, err = c.clusterClient.Update(cluster)
+			if err != nil {
+				return cluster, err
+			}
+		}
+
+		if !reflect.DeepEqual(cluster.Spec.AKSConfig, cluster.Status.AKSStatus.UpstreamSpec) {
```

Comment:
This should compare the `cluster.Spec.AKSConfig` with the `cluster.Status.AppliedSpec.AKSConfig`.

Reply:
Fixed, thank you